Make logger webhook config dynamic

anjalshireesh opened the pull request:
## Description

It should not be required to restart the server after setting the logger
webhook config.

## Motivation and Context

After enabling/disabling the logs option of callhome (that internally
configures the logger webhook), currently the system asks the user
to restart the MinIO server, which is unnecessary.

## How to test this PR?
- Build minio with this PR and start the server
- Configure a logger webhook using `mc admin config set {alias} logger_webhook` or enable/disable the logs option of callhome using `mc support callhome set {alias} logs=on|off`
- Verify that the output doesn't ask the user to restart the server

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

anjalshireesh committed:
Make logger webhook config dynamic

It should not be required to restart the server after setting the logger
webhook config.

harshavardhana commented on internal/config/config.go:
```diff
@@ -166,6 +166,7 @@ var SubSystemsDynamic = set.CreateStringSet(
 	ScannerSubSys,
 	HealSubSys,
 	SubnetSubSys,
+	LoggerWebhookSubSys,
```
can we also add `audit` here? @anjalshireesh

anjalshireesh replied:
done

anjalshireesh replied:
I've reverted it. Will work on a separate PR for making the audit webhook config dynamic.

anjalshireesh committed:
Make audit webhook config dynamic too

harshavardhana reviewed (commented):
@anjalshireesh have you tested this because? I am not sure we have code to disable an currently enabled webhook.

Because clearly from the code they are not that dynamic.

klauspost reviewed (changes requested):
This does nothing. You must handle it in `func applyDynamicConfig(ctx context.Context, objAPI ObjectLayer, s config.Config) error` as well and *safely* update the affected systems.

This must be done in a thread-safe manner, ie add or update audit targets, etc. This must work on a running system, preferably without more locks.

anjalshireesh committed:
Dynamically update logger webhook targets

anjalshireesh committed:
Revert "Make audit webhook config dynamic too"

This reverts commit cd7bf06b9383aae93c42a4e1c30c73718123c96c.

klauspost commented on cmd/config-current.go:
```diff
@@ -663,6 +663,24 @@ func applyDynamicConfig(ctx context.Context, objAPI ObjectLayer, s config.Config
 		return fmt.Errorf("Unable to apply scanner config: %w", err)
 	}
 
+	// Logger webhook
```
You can remove this from `lookupConfigs`. `applyDynamicConfig` is called at the end.

anjalshireesh replied:
Done.

anjalshireesh committed:
Remove redundent code from lookupConfigs

klauspost reviewed (approved):
LGTM

klauspost commented on cmd/config-current.go:
```diff
@@ -663,6 +651,24 @@ func applyDynamicConfig(ctx context.Context, objAPI ObjectLayer, s config.Config
+		if l.Enabled {
+			l.LogOnce = logger.LogOnceIf
+			l.UserAgent = userAgent
+			l.Transport = NewGatewayHTTPTransportWithClientCerts(l.ClientCert, l.ClientKey)
```
`l` needs to be saved back to be effective.

anjalshireesh replied:
done

harshavardhana commented on cmd/config-current.go:
```diff
@@ -663,6 +651,24 @@ func applyDynamicConfig(ctx context.Context, objAPI ObjectLayer, s config.Config
+			l.UserAgent = userAgent
+			l.Transport = NewGatewayHTTPTransportWithClientCerts(l.ClientCert, l.ClientKey)
+		}
+	}
+	err = logger.UpdateTargets(loggerCfg)
+	if err != nil {
+		logger.LogIf(ctx, fmt.Errorf("Unable to update logger webhook config: %w", err))
+	}
+
```
@anjalshireesh existing targets that are running are not canceled here so - you need to add that as well

klauspost replied:
@harshavardhana Aren't they replaced?

harshavardhana replied:
> @harshavardhana Aren't they replaced?

They run in a goroutine - I don't remember ever writing a canceler.

anjalshireesh replied:
@harshavardhana I've tried to implement canceler for the logger webhook targets. PTAL.

anjalshireesh committed:
Save config after modifying it

anjalshireesh committed:
Add canceler for logger webhook targets

klauspost commented on internal/logger/target/http/http.go:
```diff
@@ -121,54 +124,62 @@ func acceptedResponseStatusCode(code int) bool {
 }
 
 func (h *Target) startHTTPLogger() {
+	ctx, cancelFn := context.WithCancel(context.Background())
```
Doing it like this is leaky, meaning queued entries will be dropped and never sent. I will see if I can make it non-leaky.

anjalshireesh replied:
Have made changes as suggested.

klauspost commented on internal/logger/target/http/http.go:
```diff
@@ -121,54 +124,62 @@ func acceptedResponseStatusCode(code int) bool {
 }
 
 func (h *Target) startHTTPLogger() {
+	ctx, cancelFn := context.WithCancel(context.Background())
```
```Go
type Target struct {
	status int32

	// Channel of log entries
	logCh chan interface{}

	config Config
}
```

... at end of `func (h *Target) Init() error {`

```Go
	h.status = 1
	go h.startHTTPLogger()
	return nil
}
```

```Go
// Send log message 'e' to http target.
func (h *Target) Send(entry interface{}, errKind string) error {
	if atomic.LoadInt32(&h.status) == 0 {
		// Channel was closed or used before init.
		return nil
	}
	select {
	case h.logCh <- entry:
	default:
		// log channel is full, do not wait and return
		// an error immediately to the caller
		return errors.New("log buffer full")
	}

	return nil
}
```

```Go
// Cancel - cancels the target
func (h *Target) Cancel() {
	if atomic.CompareAndSwapInt32(&h.status, 1, 0) {
		close(h.logCh)
	}
}
```

You can also remove the `	go func() {` from `startHTTPLogger`, since it is already in a goroutine when started.

klauspost replied:
The only caveat about this is that events will be sent out-of-order (old target may send after new one starts)

If you do not want that, add a `sync.WaitGroup` to `Target` and have `Cancel()` wait for everything to be flushed.

anjalshireesh replied:
Thanks @klauspost ! Have made the changes suggested by you, and also added `sync.WaitGroup`. PTAL.

anjalshireesh committed:
Ensure queued entries are not dropped

and that all ongoing webhook calls are completed before new config
replaces the old one.

harshavardhana commented on internal/logger/targets.go:
```diff
@@ -108,3 +111,37 @@ func AddTarget(t Target) error {
+
+// UpdateTargets swaps targets with newly loaded ones from the cfg
+func UpdateTargets(cfg Config) error {
+	cancelAllTargets() // cancel running targets
```
why should you cancel all targets? shouldn't it be just the one that is being modified?

anjalshireesh replied:
This cancels only the `logger webhook` targets.

harshavardhana commented on internal/logger/target/kafka/kafka.go:
```diff
@@ -197,6 +197,10 @@ func (h *Target) Init() error {
 	return nil
 }
 
+// Cancel - cancels the target
+func (h *Target) Cancel() {
+}
+
```
Leaving this empty potentially causes problems, you may have to refactor the Kafka part as well since you are making the entire logger subsystem dynamic.

anjalshireesh replied:
This function is currently not called at all, so should not cause problems. But it is required as it is being added to the `Target` interface. Code can be added to it when making the audit logging config dynamic. Can be done in a separate PR.

klauspost commented on internal/logger/targets.go:
```diff
@@ -108,3 +111,37 @@ func AddTarget(t Target) error {
+
+// UpdateTargets swaps targets with newly loaded ones from the cfg
+func UpdateTargets(cfg Config) error {
+	cancelAllTargets() // cancel running targets
```
Having it here, will drop messages from when this starts until `targets` has been updated.

anjalshireesh replied:
Thanks, fixed.

klauspost commented on internal/logger/targets.go:
```diff
@@ -108,3 +111,37 @@ func AddTarget(t Target) error {
+	swapMu.Lock()
+	targets = updated
+	atomic.StoreInt32(&nTargets, int32(len(updated)))
+	swapMu.Unlock()
```
```suggestion
	swapMu.Lock()
	atomic.StoreInt32(&nTargets, int32(len(updated)))
	cancelAllTargets(targets)
	targets = updated
	swapMu.Unlock()
```

Change `cancelAllTargets` to:
```Go
func cancelAllTargets(targets []Target) {
	for _, tgt := range targets {
		tgt.Cancel()
	}
}
```
AFAICT this should ensure that we have consistent logging, meaning we don't drop and and we shouldn't have duplicates.

anjalshireesh replied:
Done

anjalshireesh committed:
Cancel the targets after acquiring lock

to avoid dropping messages

klauspost commented on internal/logger/targets.go:
```diff
@@ -108,3 +111,37 @@ func AddTarget(t Target) error {
+	cancelAllTargets() // cancel running targets
+	targets = updated
+	atomic.StoreInt32(&nTargets, int32(len(updated)))
+	swapMu.Unlock()
```
I has the order quite specific, updating `nTargets` as the first thing, since this is used outside the lock. 

I am not sure it matters, since only "zero" or non-zero matters. Going TO zero will cause targets to be empty, so no big deal.

Going *from* zero will make it take a bit longer, but I don't suppose that matters.

I think this will work. Just a bit less sure.

anjalshireesh replied:
I felt it made sense to update the `nTargets` after updating the targets. Nevertheless, have changed the code to what you suggested.

anjalshireesh committed:
Update nTargets before cancelling targets

klauspost reviewed (approved):
Thanks for changing it. It is rather complex, so hard to get right. This should yield the correct results AFAICT. At least this is the one I am most confident with.
